File: LeetCode/120. Triangle/Solution.cpp
#include <iostream>
#include <vector>

using namespace std;

/**
 * 	Time Limit Exceed
 */
class TLESolution {
public:
    int minimumTotal(vector<vector<int>>& triangle, int row, int col) {
        if (row == triangle.size() - 1) return triangle[row][col];
        return triangle[row][col] + 
        min(minimumTotal(triangle, row + 1, col), minimumTotal(triangle, row + 1, col + 1));
    }
    int minimumTotal(vector<vector<int>>& triangle) {
        return minimumTotal(triangle, 0, 0);
    }
};

/**
 * 43 / 43 test case passed
 * Status: Accepted
 * Runtime: 8 ms
 */
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        for (int i = triangle.size() - 2; i >= 0; -- i) {
            for (int j = 0; j < triangle[i].size(); ++ j) {
                triangle[i][j] += min(triangle[i + 1][j], triangle[i + 1][j + 1]);
            }
        }
        return triangle[0][0];
    }
};

int main () {
    vector<vector<int>> triangle;
    int data[4][4] = {
        {2},
        {3, 4},
        {6, 5, 7},
        {4, 1, 8, 3}};
    for (int i = 0; i < 4; ++ i) {
        vector<int> row;
        for (int j = 0; j <= i; ++ j) {
            row.push_back(data[i][j]);
        }
        triangle.push_back(row);
    }
    cout << Solution().minimumTotal(triangle) << endl;
}

File: LeetCode/120. Triangle/Solution.kt
/**
 * Created by Inno Fang on 2017/12/2.
 */

/**
 * 43 / 43 test cases passed.
 * Status: Accepted
 * Runtime: 336 ms
 */
class Solution {
    fun minimumTotal(triangle: List<List<Int>>): Int {
        val tri = triangle as List<MutableList<Int>>
        (tri.lastIndex - 1 downTo 0).forEach { i ->
            (0 until tri[i].size).forEach { j ->
                tri[i][j] += minOf(tri[i + 1][j], tri[i + 1][j + 1])
            }
        }
        return tri[0][0]
    }
}


fun main(args: Array<String>) {
    val triangle = listOf(
            arrayListOf(2),
            arrayListOf(3, 4),
            arrayListOf(6, 5, 7),
            arrayListOf(4, 1, 8, 3))

    Solution().minimumTotal(triangle).let(::println)
}